pkg/informer/informer.go

```diff
@@ -454,6 +457,56 @@ func (d *DynamicDiscoverySharedInformerFactory) updateInformers() {
 		delete(d.informerStops, gvr)
 		delete(d.startedInformers, gvr)
 	}
+
+	d.discoveryData = gvrsToDiscoveryData(latest)
+}
+
+func gvrsToDiscoveryData(gvrs map[schema.GroupVersionResource]struct{}) []*metav1.APIResourceList {
+	var discoveryData []*metav1.APIResourceList
+	gvResources := make(map[schema.GroupVersion][]metav1.APIResource)
+
+	for gvr := range gvrs {
+		gv := gvr.GroupVersion()
+
+		gvResources[gv] = append(gvResources[gv], metav1.APIResource{
+			Name:    gvr.Resource,
+			Group:   gvr.Group,
+			Version: gvr.Version,
+			// Everything we're informing on supports these
+			Verbs: []string{"create", "list", "watch", "delete"},
+		})
+	}
+
+	for gv, resources := range gvResources {
+		sort.Slice(resources, func(i, j int) bool {
+			return resources[i].Name < resources[j].Name
+		})
+
+		discoveryData = append(discoveryData, &metav1.APIResourceList{
+			GroupVersion: gv.String(),
+			APIResources: resources,
+		})
+	}
+
+	sort.Slice(discoveryData, func(i, j int) bool {
+		return discoveryData[i].GroupVersion < discoveryData[j].GroupVersion
+	})
+
+	return discoveryData
+}
+
+// DiscoveryData implements resourcequota.NamespacedResourcesFunc and is intended to be used by the quota subsystem.
+func (d *DynamicDiscoverySharedInformerFactory) DiscoveryData() ([]*metav1.APIResourceList, error) {
+	d.informersLock.RLock()
+
+	ret := make([]*metav1.APIResourceList, len(d.discoveryData))
+	for i, apiResourceList := range d.discoveryData {
+		ret[i] = apiResourceList.DeepCopy()
+	}
+
+	d.informersLock.RUnlock()
```

Comment:
why not the defer variant if it is possible? Seems to be simpler and safer.

Reply:
I remember Steve saying it's faster without the defer, but I'll switch it in the next push